### ISSUE TITLE
DWA in tensorflow

### ISSUE THREAD

**Issue by Zxlan** [lorenmt/mtan#2]
Have you tried to use tensorflow to complete the DWA module?

**Comment by lorenmt** [comment]
All of my implementations were evaluated in pytorch. And I believe to apply DWA in tensorflow should be quite simple and straightforward. If you planned to reimplement DWA module in tensorflow, I may assist or have a final review.

**Comment by Zxlan** [comment]
I want to learn from the idea of DWA, but in the implementation of tensor operation encountered a lot of errors. So I want to ask you if you use tensorflow to implement it.




------------------ 原始邮件 ------------------
发件人: "Shikun Liu"<notifications@github.com>;
发送时间: 2019年5月26日(星期天) 晚上8:54
收件人: "lorenmt/mtan"<mtan@noreply.github.com>;
抄送: "瑶望&兰馨"<1069081791@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [lorenmt/mtan] DWA in tensorflow (#2)




All of my implementations were evaluated in pytorch. And I believe to apply DWA in tensorflow should be quite simple and straightforward. If you planned to reimplement DWA module in tensorflow, I may assist or have a final review.

**Comment by lorenmt** [comment]
DWA is only a learning strategy in multi-task learning to update task-weighting over time. Thus the weighting should be in scalar rather than tensor. If you met errors in reimplementation, I suggest it's better to check the tensorflow official documentations or tutorials. Hope that helps.

**Comment by Zxlan** [comment]
The weight is calculated from the previous epoch loss. How can it be scalar? I don't know.




------------------ 原始邮件 ------------------
发件人: "Shikun Liu"<notifications@github.com>;
发送时间: 2019年5月26日(星期天) 晚上9:07
收件人: "lorenmt/mtan"<mtan@noreply.github.com>;
抄送: "瑶望&兰馨"<1069081791@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [lorenmt/mtan] DWA in tensorflow (#2)




DWA is only a learning strategy in multi-task learning to update task-weighting over time. Thus the weighting should be in scalar rather than tensor. If you met errors in reimplementation, I suggest it's better to check the tensorflow official documentations or tutorials. Hope that helps.

**Comment by Zxlan** [comment]
Tensorflow builds the graph first and then trains. This DWA seems to change the graph during the training process. So I'm confused.




------------------ 原始邮件 ------------------
发件人: "Shikun Liu"<notifications@github.com>;
发送时间: 2019年5月26日(星期天) 晚上9:07
收件人: "lorenmt/mtan"<mtan@noreply.github.com>;
抄送: "瑶望&兰馨"<1069081791@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [lorenmt/mtan] DWA in tensorflow (#2)




DWA is only a learning strategy in multi-task learning to update task-weighting over time. Thus the weighting should be in scalar rather than tensor. If you met errors in reimplementation, I suggest it's better to check the tensorflow official documentations or tutorials. Hope that helps.

**Comment by lorenmt** [comment]
No. DWA does not change the graph at all. It only provides a scalar to renonrmalise/balance the gradient from each task loss. Please recheck the description from our paper.

**Comment by Zxlan** [comment]
Thank you very much.I made it.It turned out that I was wrong and complicated it.




------------------ 原始邮件 ------------------
发件人: "Shikun Liu"<notifications@github.com>;
发送时间: 2019年5月27日(星期一) 中午1:22
收件人: "lorenmt/mtan"<mtan@noreply.github.com>;
抄送: "瑶望&兰馨"<1069081791@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [lorenmt/mtan] DWA in tensorflow (#2)




No. DWA does not change the graph at all. It only provides a scalar to renonrmalise/balance the gradient from each task loss. Please recheck the description from our paper.

**Comment by hu-zhiyu** [comment]
could you please share your DWA implementation in tensorflow with me? Thanks a lot.

**Comment by jeffbaena** [comment]
Dear authors, would it be possible to explain which value K has been choosen  and why?

![Screen Shot 2020-01-21 at 13 39 55](https://user-images.githubusercontent.com/18122010/72805576-86d0bb80-3c53-11ea-8bdd-596e924ef09d.png)

**Comment by lorenmt** [comment]
Hi,

Index K simply represents the index of the task. Thus k is not chosen.
Please directly check out the source code for better understanding.

Best,
Sk.

**Comment by jeffbaena** [comment]
Perfect, thanks for the quick reply!!

**Comment by lorenmt** [comment]
Sorry. If you meant the capital K, it represents the total number of tasks, i.e. If we are doing two-task multi-task learning, K would be 2.

And then smaller k would be 1 or 2 (task specific index).
The capital K is performing as a normalisation factor to make sure
sum_k lambda_k = K, which gives the similar gradient magnitude from equal weighting.

Hope that clears out your question/
Sk.

**Comment by jeffbaena** [comment]
Yes this is more clear to me! Thanks

**Comment by jeffbaena** [comment]
Dear @lorenmt thanks for your help. I would really like to use your DWA algorithm to add a loss function for Optical Flow estimation. The additional loss function (loss IMBALANCE) intuitively should help the main loss (EPE loss). I have implemented your algorithm to update every 280 iterations, with T= 2.

However, I see that the lambda coefficients strongly oscillates around 1, but the two losses (by construction) have different scales so I expect the calculated lambdas to be very different (and more stable over training time).

The first two plots are the calculated lambdas (using DWA).
The second ones are the original losses (not multiplied by lambda)

Any hint would be greatly appreciated.

Regards,
Stefano

![Screen Shot 2020-01-22 at 11 31 28](https://user-images.githubusercontent.com/18122010/72886810-c6a6aa00-3d0a-11ea-9a1c-168cd308b95d.png)
![Screen Shot 2020-01-22 at 11 31 24](https://user-images.githubusercontent.com/18122010/72886812-c73f4080-3d0a-11ea-95ff-c5de6f22b8b4.png)
![Screen Shot 2020-01-22 at 11 31 15](https://user-images.githubusercontent.com/18122010/72886814-c73f4080-3d0a-11ea-9134-d313ebada17a.png)

**Comment by lorenmt** [comment]
Hi Stefano,

Thanks for your question.

From the graphs you provided, they seem that the EPE and IMBALANCE loss are significantly different in terms of scale. 

Assuming you are performing image-to-image type of tasks, I am wondering whether you are able to rescale the loss function, say what I did for semantic and depth prediction is to normalize across per pixel, so that the scale would not have to become much difference.

If you really cannot modify the loss function, I suggest you increase the temperature T to make it more steady, say you can try T = 10 or 20.

Hope that helps.
Sk.

**Comment by jeffbaena** [comment]
All right! I will try to rescale the loss, I will start just by multiplying the loss by a constant and see how it goes.

Thanks so much,
Stefano
